черновик

cirill98 opened the pull request.

cirill98 committed:
черновик

cirill98 committed:
контроллер и сервис прописаны, настроил в другом микросервисе проверку User. Протестировал через Postman и через Юнит и Мок тесты

cirill98 committed:
Внес корректировки

cirill98 committed:
Внес корректировки

cirill98 committed:
Внес корректировки

raukaztaev commented on src/test/java/faang/school/postservice/service/comment/CommentServiceTest.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    private void testValidateAuthorExists() {
+        Long postId = 1L;
```
postId лучше вынести в поле класса, а то ты его создаешь в каждом тесте

itstnslv commented on src/main/java/faang/school/postservice/controller/ApiPath.java:
```diff
@@ -0,0 +1,9 @@
+package faang.school.postservice.controller;
+
+public final class ApiPath {
+    public static final String COMMENT = "/comment";
```
Будь готов к первому мердж конфликту :) либо я буду готовиться, в зависимости от того, кто первый смерджится

itstnslv commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,42 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(ApiPath.COMMENT)
```
Для всех последующих эндпоинтов можно в константы тоже закинуть плейсхолдер с /{id}. Глаголы можно убрать, потому что они уже описаны в HTTP методе.

itstnslv commented on src/main/java/faang/school/postservice/validation/dto/DtoValidation.java:
```diff
@@ -0,0 +1,7 @@
+package faang.school.postservice.validation.dto;
+
+public final class DtoValidation {
```
Здесь нам тоже стоит прийти к одному названию DtoValidationConstraints, чтобы просто разрешить конфликт и объединить все константы вместо того, чтобы плодить два отдельных класса.

itstnslv commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,75 @@
+        UserDto user = userServiceClient.getUser(authorId);
+        if (user == null) {
+            log.error("Пользователь с таким ID не найден");
+            throw new NullPointerException("Пользователь с таким ID не найден");
```
Уместней будет если мы в таких случаях будем выбирать EntityNotFoundException

itstnslv commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,75 @@
+        Comment comment = commentMapper.toEntity(commentDto);
+        comment.setPost(postService.getPost(id));
+        comment.setLikes(new ArrayList<>());
+        commentRepository.save(comment);
```
Комменты могут сыпаться с большой скоростью. Полезно будет, если мы предусмотрим здесь обработку ошибок с просьбой повторить попытку написания коммента или указанием на то, что пользователь делает неправильно.

itstnslv commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,75 @@
+    private final PostService postService;
+    private final CommentMapper commentMapper;
+
+    public CommentDto addCommentService(Long id, CommentDto commentDto) {
```
Названия методов вводят в смятение немного. Сервис тут уже не при делах, нам стоит описывать логику, которая происходит с комментами

itstnslv commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -0,0 +1,21 @@
+    public Post getPost(long id) {
+        return postRepository.findById(id).orElseThrow(() -> {
+            log.error("Поста по такому id не существует");
+            return new NullPointerException("Поста по такому id не существует");
```
Тоже обрати внимание на тип исключения

cirill98 replied:
Тесты у меня ожидают именно NullPointerException)) надо логику переписывать, чтобы другое исключение выпало

egorbarzenkov commented on src/main/java/faang/school/postservice/controller/comment/CommentController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @PutMapping("/update/{id}")
+    public ResponseEntity<CommentDto> updateCommentController(@PathVariable Long id,
```
не обязательно называть методы "Controller"
так же обрати внимание на поле Long id, это на сколько я понимаю идентификатор поста, такое же поле у тебя есть в CommentDto, postId либо оставь одно из них, либо добавь проверку на их равенство, но лучше как мне кажется иметь один идентификатор в одном месте

egorbarzenkov commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,75 @@
+        return commentMapper.toDto(comment);
+    }
+
+    public List<CommentDto> getCommentsService(Long id) {
```
getCommentsForPost

egorbarzenkov commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,75 @@
+
+    public CommentDto deleteCommentService(Long id, CommentDto commentDto) {
+        validateAuthorExists(commentDto.getAuthorId());
+        Comment comments = findCommentInPost(id, commentDto);
```
можно добавить условие, что так мы ищем комент только если в commentDto отсутствует id
так же сюда можно докинуть проверку "на дурака", что если у нас есть commentDto.id и commentDto.postId то соответственно проверять, что комментарий с таким идентификатором есть у поста, указанного в dto

cirill98 replied:
"можно добавить условие, что так мы ищем комент только если в commentDto отсутствует id" у меня по валидации поле postId в CommentDto не может быть пустым, а id я здесь поставил, потому что по условию он должен быть в пар-ре метода

egorbarzenkov commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,75 @@
+                .orElseThrow(() -> new IllegalArgumentException("Комментарий с таким содержанием и автором не найден"));
+    }
+
+    private void validateAuthorExists(long authorId) {
```
так же на будущее, лучше разбивать проверку пользователя и выбрасывание исключения на разные уровни, т.е. отдельно вынести сервис проверки юзера, и уже отдельно сервис отвечающий за генерацию исключений, потому что в дальнейшем может понадобится логика без прерывания операции и тогда придется пилить её через try catch блоки, что не способствует упрощению логики

cirill98 replied:
Метод нужно вынести в класс валидации в отдельном пакете? И нужен класс для общих исключений?

egorbarzenkov replied:
да, в отдельный пакет, можно положить рядом с UserServiceClient, можно создать исключение с текстовкой "user {id} not found in user service", которое будет использоваться только при неуспешном получении пользователя

egorbarzenkov replied:
так, ты вынес валидатор в отдельный объект, но все равно оставил валидацию в сервисе, недоделка получается?

cirill98 committed:
Внес корректировки

cirill98 committed:
Внес корректировки

cirill98 committed:
Внес корректировки

cirill98 committed:
1) Добавил пакет с исключениями и класс с сообщениями об исключении
2) Добавил проверку в addNewCommentInPost, чтобы в случае ошибки повторили запрос

egorbarzenkov commented on src/main/java/faang/school/postservice/service/post/PostService.java:
```diff
@@ -0,0 +1,21 @@
+    public Post getPost(long id) {
+        return postRepository.findById(id).orElseThrow(() -> {
+            log.error("Поста по такому id не существует");
+            return new NullPointerException("Поста по такому id не существует");
```
поправь тип исключения, NullPointerException лучше не использовать в такой логике, можно лучше EntityNotFoundException

egorbarzenkov replied:
так же обрати внимание на то, что не все серверы могут поддерживать разные кодировки, по этому русского языка стоит в ошибках избегать

egorbarzenkov commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,89 @@
+    private void validateAuthorExists(long authorId) {
+        UserDto user = userServiceClient.getUser(authorId);
+        if (user == null) {
+            log.error(ExceptionMessages.POST_NOT_FOUND);
```
почему проверяется юзер а выбрасывается POST_NOT_FOUND?

egorbarzenkov replied:
так же вынеси эту проверку в отдельный сервис, скорее всего её часто будут переиспользовать

cirill98 replied:
В какой сервис вынести? создать отдельный класс валидации и туда этот метод положить?

egorbarzenkov commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,89 @@
+
+    public CommentDto deleteExistingCommentInPost(CommentDto commentDto) {
+        Comment comments = findCommentInPost(commentDto);
+        commentRepository.deleteById(comments.getId());
```
у тебя же в dto передается идентификатор комментария, зачем искать его через пост и потом еще сравнивать по контенту?

cirill98 replied:
логика была, чтобы провести валидацию. т.е. быть уверенным, что мы удаляем нужный коммент. Потом просто перевожу обратно в ДТО и возвращаю.
Если я без валидации, просто буду удалять по id, то что метод будет возвращать?

egorbarzenkov replied:
это уже по сути не твоя проблема, пользователь должен решать какой комментарий он хочет удалить, искать его по посту и проверять по контенту не имеет смысла, очень много лишних действий происходит, идея возвращать dto удаленного комментария в целом хороша, но получать его лучше по идентификатору, упрости логику и будет хорошо

egorbarzenkov commented on src/main/java/faang/school/postservice/service/comment/CommentService.java:
```diff
@@ -0,0 +1,89 @@
+
+    public CommentDto updateExistingComment(CommentDto commentDto) {
+        validateAuthorExists(commentDto.getAuthorId());
+        Comment comment = commentRepository.findAllByPostId(commentDto.getPostId()).stream()
```
у тебя в commentDto есть id комментария, можно получать комент по нему и уже его изменять если такой есть, слишком переусложнил логику

cirill98 committed:
упростил логику

egorbarzenkov commented on src/main/java/faang/school/postservice/validation/comment/AuthorExistsValidation.java:
```diff
@@ -0,0 +1,26 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class AuthorExistsValidation implements CommentValidator{
```
почему AuthorExistsValidation имплементирует  CommentValidator?
как то не логично получается
Так же проблема в том, что у валидаторов может быть разная логика валидации бизнес параметров и соответственно разные поля под это бизнес поведение и либо всем придется имплементировать один метод который принимает только объект CommentDto, (что еще одно замечание, в выделенном функционале отсутствует гибкость имплиментации), и выделять под разные сценарии разные валидаторы, либо все забьют на это и будут пилить валидаторы без интерфейса

cirill98 committed:
1) добавил свое исключение
2) сделал отдельный класс-валидацию
